### PR TITLE
Update SFML version to 3.0.0

### PR DESCRIPTION
**Pull request by eXpl0it3r**
## Description

Quite some while ago, there was an internal discussion and vote to bump release version numbers right after a release or in this case, when creating the "release branch".

Unlike with SFML 2.x and earlier the version numbers are updated as soon as work on the new release starts. This especially helps with version checks, which until now caused issues with in-development version matching the previous release.

### CODE REVIEW

**Comment by Bromeon** [comment]
> Unlike with SFML 2.x and earlier the version numbers are updated as soon as work on the new release starts. This especially helps with version checks, which until now caused issues with in-development version matching the previous release.

Doesn't this just shift the problem?

Now, an in-dev 3.0.0 version will be incompatible with the 3.0.0 release.
So, a preprocessor check for 3.0.0 is _still_ unreliable.

This is an even bigger issue for 3.0.0 specifically, since SFML 3 is expected to be quite a while in development*, as it's the only chance to introduce breaking changes for the forseeable future. In my opinion, it would be better to set it to something like 2.7 or so, to make clear it doesn't match a released version.

___
\* I know the plan is to iterate quickly on SFML 3, but we should also keep longer-term stability in mind. No one is helped if SFML 4 is released half a year after SFML 3, it just means people have to break their code multiple times.

**Comment by eXpl0it3r** [comment]
I think the [past discussion](https://en.sfml-dev.org/forums/index.php?topic=20278.0) (internal-only) covered most of these concerns. It won't really shift the problem, if people rely on unreleased versions, then they accept the risk of breaking changes.

There was also a discussion for adding a macro like `SFML_VERSION_UNRELEASED` which many agreed on, but afterwards there was some strong opinions against it. For me both works. We don't *really* have to know whether it's a development version or not, but I can also see situations where people could find it useful.

**Comment by Bromeon** [comment]
Thanks for the link! 👍 

> It won't really shift the problem, if people rely on unreleased versions, then they accept the risk of breaking changes.

The same is true if the unreleased SFML keeps the version of the last released one. I don't see how one is better than the other.
That said, I'm OK with both 🙂 

But maybe we should go a step back. Who is using the macros?
* Build systems (my project requires at least SFML 2.3)
* Source code to allow multiple SFML versions side-by-side (use X if SFML < 3; use Y if SFML >= 3)
* Anyone else?

---

`SFML_VERSION_UNRELEASED` seems helpful to me, but I see how it would be a bit cumbersome to change it just for the release and then revert. One option I see is to do it in CI -- have a simple `sed` statement which replaces that line for the built package artefact. Then it always stays at value `0` for the Git version.

If  we go with such a macro, we could even include the commit SHA from CI.
```cpp
#define SFML_GIT_VERSION "6cf124db66f2e6986b2fb9e18e47d9cba2799fec"
```
and remove it / set it to a special value for released versions.

**Comment by eXpl0it3r** [comment]
> The same is true if the unreleased SFML keeps the version of the last released one. I don't see how one is better than the other.

Yes and no. The reason the new solution is "better" is because the case using the latest in-development version and wanting some version checks is a much more common case, than requiring a minimum version and someone using an in-development / not yet released version.

Best example would be the new major version. If we don't change anything, then SFML 2.5.1 (or later 2.6.0) can be requested (e.g. with CMake `find_package(SFML 2.6 COMPONENTS graphics)` and CMake would happily accept the in-development SFML 3 version and then of course fail during linking, as all the APIs are changed.

If someone checks out the master branch right after merging this and then a year down line when SFML is released, tries to link this SFML "3" version, it will also fail, but this time it's a clear failure by the user, who relied on a development version from a year ago.

If you think in range notation and you check for 2.5.1 you can only check the following: `[2.1 ... 2.6.0[`
While with the new version it would essentially be `]2.6.0 ... 3.0.0]`

I don't really see the use cases for having the information of `]2.6.0 ... 3.0.0[` or `[3.0.0 ... ]` available. Plus marking something explicitly as "unreleased" can also have some negative effect on adaptation, which of course can be countered a bit with more frequent releases.

A hash doesn't really help to determine whether something is released or not and it's only useful if you can go look it up later on, which I'm not sure how far back you can/want to go on the CI. Also build versioning is a bit of a different topic. I would only do this, if also start publishing every single build to GitHub Release or somewhere else.

**Comment by Bromeon** [comment]
> Best example would be the new major version. If we don't change anything, then SFML 2.5.1 (or later 2.6.0) can be requested (e.g. with CMake `find_package(SFML 2.6 COMPONENTS graphics)` and CMake would happily accept the in-development SFML 3 version and then of course fail during linking, as all the APIs are changed.

Thanks for the elaboration! 🙂 

During a breaking change (2.6 -> 3.0), both scenarios are problematic:
1. When requesting `2.6`, I don't want to get in-dev 3.0
2. When requesting `3.0`, I don't want to get in-dev 3.0 which is basically `2.6` + some changes

But I agree that the second is a bit better given chronology.
Unfortunately there seems to be no CMake mechanism to express that -- I could only think of something like `2.7` to avoid incompatibilities. But at least then, people can rely that when they request `3.0`, they will also all or nothing.

**However**, during minor increments, I'd say the opposite is true.
1. When requesting `2.6`, I want to get **all** the features that version 2.6 offers, and definitely not 2.5 + some changes.
2. On the other hand, it does _not_ matter if I get **2.6** + some changes, since those are backwards-compatible.

So, for minor versions it's better to keep the last released version until next release.

---

> A hash doesn't really help to determine whether something is released or not and it's only useful if you can go look it up later on, which I'm not sure how far back you can/want to go on the CI.

It would be more for traceability -- people who want to know exactly which version they have. Maybe they build a tool using SFML and want `--version` to output the underlying tech stack. The other use case would be to check if someone is using the latest version.

Whether something is released or not _can_ be expressed using the hash, e.g. using a certain convention, or "version debug string":
```cpp
#define SFML_VERSION_STR "3.0.1"         // released version
#define SFML_VERSION_STR "dev-6cf124d"   // GitHub version
```
This is what many tools do, `rustc` and `godot` come to my mind.

Any such changes would only be for new versions (starting at in-dev 3.0.0), not retroactively. But I agree it's not terribly useful, so if it comes at the cost huge CI changes, let's not do it.

**Comment by eXpl0it3r** [comment]
I decided to just add a macro and CMake variable for now. If it turns out at a later point, that we don't actually need it, we can always remove it again and there's no harm done.

Adding revision magic is quite a bit more work for no real benefits at this stage.
If we ever do rolling releases, nightly builds or reproducible builds, then it's a must, but for now it's not really something we should spend time on.

File: CMakeLists.txt
Comment (kimci86) [inline CMakeLists.txt:57]: This might lead to writing double negations like `if(NOT VERSION_UNRELEASED)`.

Comment (Bromeon) [reply CMakeLists.txt:57]: Good point, maybe `VERSION_IS_RELEASE` is better. 
The "is" also makes clear it's a boolean flag and not a version, as otherwise it follows the same naming as the 3 versions.

Comment (eXpl0it3r) [reply CMakeLists.txt]: Like this? 🙂

**Comment by Bromeon** [comment]
Thanks, adapations look mostly good! 🙂

Only one topic:

> During a breaking change (2.6 -> 3.0), both scenarios are problematic:
> 1. When requesting `2.6`, I don't want to get in-dev 3.0
> 2. When requesting `3.0`, I don't want to get in-dev 3.0 which is basically `2.6` + some changes

Using `2.7` would solve this. It's unconventional, however from a practical perspective it's more reliable.

Keep in mind that once we release 3.0.0, we will want to keep both macro and CMake variable at that version, since in-between changes are backwards-compatible but not forward-compatible (see reasoning outlined above). So we'll have 3.0.0 to denote anything *before*, *at* and *after* the actual SFML 3 release, which makes it not very useful.

**Comment by eXpl0it3r** [comment]
> Using `2.7` would solve this. It's unconventional, however from a practical perspective it's more reliable.

Referencing a version that will never exist, is problematic on its own. 😄 
Anyways, I don't want to start any additional discussions, especially since we're mostly discussing very edge cases scenarios with no practical evidence. 😉

Now, I'm just waiting for a review approval.

**Comment by Bromeon** [comment]
> Anyways, I don't want to start any additional discussions, especially since we're mostly discussing very edge cases scenarios with no practical evidence. 😉

Using CMake `find_package()` with 3.0.0 **will not return v3.0.0**, or even a compatible one.
Not exactly an edge case.

If our position is "that's the user's problem", OK.

**Comment by eXpl0it3r** [comment]
Since that didn't lead to an approval, what do you think is still open?

**Comment by eXpl0it3r** [comment]
Do you mean `SFML_STATIC_LIBRARIES`?

I'm still convinced that we're discussing scenarios that in practice will be a non-issue.

**Comment by Bromeon** [comment]
Since we now expose the `SFML_VERSION_IS_RELEASE` flag, the user can check on his own if he insists on a released version, so I don't think we should add an extra flag when calling `find_package`.

The only thing he can't do is know whether the source _compatible_ to SFML 3.0.0, that is released afterwards or before, as those two will look the same from a version's point of view.

As mentioned, I'd be OK with multiple options:
* Something like `2.7` or rather `2.9.9` to make clear it's "not yet 3.0.0" and breaking changes are still possible. This would be the easiest and maintain chronology.
* Another flag `SFML_VERSION_UNSTABLE` indicating that there are still breaking changes, this would be removed once we get to 3.0.0.
* Do nothing but precisely document the behavior.
  If it turns out to be a real issue, we could still add `SFML_VERSION_STABLE` after 3.0.0.

**Comment by eXpl0it3r** [comment]
So what's the point of `SFML_VERSION_IS_RELEASE` if you also need `SFML_VERSION_UNSTABLE`.

I'm honestly *very* annoyed that we can't make progress here and block the whole branch...

**Comment by kimci86** [comment]
> Do you mean SFML_STATIC_LIBRARIES?

Yes. For example, if only shared libraries are installed on a system, this will fail:
```cmake
set(SFML_STATIC_LIBRARIES TRUE)
find_package(SFML 2.5 COMPONENTS graphics REQUIRED)
```

I thought there could the same behavior for released/in-development versions:
```cmake
find_package(SFML 3.0 COMPONENTS graphics REQUIRED) # fail

set(SFML_PLEASE_LET_ME_USE_IN_DEVELOPMENT_VERSION TRUE)
find_package(SFML 3.0 COMPONENTS graphics REQUIRED) # success
```

But no worries, it was just some idea.

**Comment by Bromeon** [comment]
> So what's the point of `SFML_VERSION_IS_RELEASE` if you also need `SFML_VERSION_UNSTABLE`?

**It's not possible to differentiate `< 3.0.0` or `> 3.0.0` with the suggested resolution of this PR.**

This flag would be one of many suggestions to make this possible (`UNSTABLE` meaning `< 3.0.0`).
I'm open to alternative suggestions that allow this, I really don't care how it's done.

---

>  And again, no, I won't be accepting fake versions for "fake problems".
> 
> I'm honestly very annoyed that we can't make progress here and block the whole branch...
> Up to the point of seriously thinking about abandoning this PR and just continuing with the worse model.

I gave multiple good reasons why I think the new versioning model is still flawed. You never addressed them beyond "it's not a real problem" -- which could be said about this PR as well, and still we're trying to improve on versioning, no?

It's totally OK if you don't agree with the suggestions -- I even offered twice to simply document the behavior.

But saying I'm creating "fake problems" and deliberately blocking progress is not constructive. Especially when I have recently invested much more time for SFML reviews than usual, in a hope to get things back to speed and unblock the C++11 progress.

**Comment by eXpl0it3r** [comment]
> This flag would be one of many suggestions to make this possible (UNSTABLE meaning < 3.0.0).

If it's not a release (`SFML_VERSION_IS_RELEASE` = false) version 3.0.0 and you can check that exactly, or what am I missing?
What's the use case for `SFML_VERSION_IS_RELEASE` then? Because I added this to solve exactly this issue...

**Comment by Bromeon** [comment]
> If it's version 3.0.0 and **not** a release (`SFML_VERSION_IS_RELEASE` = false) then it's an unstable 3.0.0 version.

No, it could also be 3.1.0 in development.
In this case it _**is**_ backwards-compatible with 3.0.0, so a user project that needs 3.0.0 could depend on it without problems.

As mentioned, we can go ahead and document it. I can provide some documentation if it helps.

**Comment by eXpl0it3r** [comment]
> No, it could also be 3.1.0 in development.
> In this case it is backwards-compatible with 3.0.0, so a user project that needs 3.0.0 could depend on it without problems.

I don't understand what you mean here. If you do `find_package(SFML 3.0 COMPONENTS graphics)` and then check that `VERSION_IS_RELEASE` is set, you a) accept only true releases (so no in-development version) and b) if CMake finds SFML 3.1 it will accept that as well, since it's backwards-compatible. 
I don't see any gap that `SFML_VERSION_UNSTABLE` would fill.

**Comment by eXpl0it3r** [comment]
Okay, yes it was not clear to me, that your argument started again at the beginning and isn't based on the applied changes 😄 

But your argument falls apart for breaking changes, because the in-dev version is only backwards-compatible up to the point where the breaking change is introduced. At that point you have a really fun time of I want version 2.6. and you get the development version with all the braking C++17 changes (for example).

The reason I still don't see this as a critical issue is because:

1. The user who goes and defines "I want SFML 2.6" while SFML 2.6 is still in development, knows exactly what they sign up for. They want to signal to anyone else using their code, that they depend on features that are not available in SFML 2.5. In case the development version somehow breaking their implementation, then they have to go and fix it, but that's the risk of using an in-dev version.
2. Now once SFML 2.6 is released, the user who has somehow an in-dev version laying around or downloaded it from somewhere and tries to use the latest 2.6 feature, not available in their in-dev version gets a breaking build. I consider this an edge case, because in-dev versions don't magically end up on your system, so by whatever means you got a bad/outdated set of libs, you can also go and get the latest and properly released ones.

With currently implemented approach:

- You don't run into the issue, where a code base pretends to be an already released version, but with new features, bugfixes or even breaking changes
- You can clearly define intend of "I want to use features of the next version"
- You can support multiple versions, including ones that are still in development
- You can check whether you're on an actually released version or something in-between

And a fake version won't solve any of this, because at the end of the day, you can also request a fake version and run into all the same issues. 😉 

---

> I can provide some documentation if it helps.

If you have anything specifically in mind, would be great if you can provide it. 🙂

**Comment by Bromeon** [comment]
Thanks a lot for the detailed answer! 🙂 

> But your argument falls apart for breaking changes, because the in-dev version is only backwards-compatible up to the point where the breaking change is introduced.

That's true, and it's why I was looking for (admittedly not very beautiful) alternatives like `UNSTABLE` or `2.9.9`. The only difference it would make is that it clearly states "I am neither 2.6 nor 3.0".

---

From your explanation, it seems like you have a bit a different approach than me in mind. Mine is roughly:
* If you specify `3.1`, you will get a version that's **always** gonna be backwards-compatible to `3.1`, so it must be either the release or a future version.

If I understood your way, it's this:
* Someone specifying version `3.1` may also want to use `in-dev 3.1`. They will always get the latest version, but this may not be the full feature set of the final `3.1` release.
* If someone wants to rely on semver compatibility, he should stick to released versions only.

I think this is a legitimate approach, if communicated accordingly. The only risk I see that someone who has several versions (or gets them automatically through Git or so) may accidentally use an older version than the one requested, because they're not aware of `in-dev` versions implicitly found. This is where @kimci86's suggestion would come into play, but doc could also do the trick.

---

Regarding docs, what do you think of this?

> When searching for SFML with find_package(), keep in mind that it will also find versions which are in development (i.e. between two released versions), if you have them in your search path. If you want to make sure that a found SFML package corresponds to an official release, check the bool output variable VERSION_IS_RELEASE, which is true for official releases and false for development versions.
> 
> If you want to use the latest features before a new SFML version is released, make sure to look for the *upcoming* version number. For example, you want to use latest SFML 3 features before 3.0.0 it is officially released, look for version 3 with find_package().

This raised a question though: how do we know whether after 3.0.1 there will be 3.0.2 or 3.1.0? Or what should a user request if he wants the latest features?

Or do you think that once SFML 3 is released, we should still keep the previous version until the next release (instead of the upcoming one)?

**Comment by eXpl0it3r** [comment]
> Regarding docs, what do you think of this?

Sounds great, thanks for writing it! 🙂 

> This raised a question though: how do we know whether after 3.0.1 there will be 3.0.2 or 3.1.0? Or what should a user request if he wants the latest features?

Patch releases are essentially "hotfixes" and we usually don't plan them, thus the next version after 3.0.0 will be 3.1.0.
If we notice that there some severe bugs in 3.0.0, we can branch off from the release and introduce a 3.0.1 version in between with the necessary fixes.

**Review by Bromeon — APPROVED** [review]

**Comment by Bromeon** [comment]
Good point; I was somehow thinking in libraries with `0.6.1`, where the 3rd digit is often used as a "minor" version, meaning new features.

For me, the changes are good to go, thanks for taking care of this PR! 🙂

File: cmake/SFMLConfig.cmake.in
Comment (kimci86) [inline cmake/SFMLConfig.cmake.in:34]: Typos here. This should be:
> For example, if you want to use latest SFML 3 features before 3.0.0 is officially released, look for version 3 with find_package().

Comment (eXpl0it3r) [reply cmake/SFMLConfig.cmake.in:34]: Thanks! 🙂

**Comment by kimci86** [comment]
SFMLConfig.cmake.in should be updated to define VERSION_IS_RELEASE somehow, or did I miss something ?

**Comment by eXpl0it3r** [comment]
It's in in the CMakeLists.txt itself, where the rest of the version is defined.
If you mean, that there should be a check or anything, that's left to the consumer to decide.

What's your expectation of `VERSION_IS_RELEASE` in CMakeConfig.cmake.in?

**Comment by kimci86** [comment]
When using `find_package` to find SFML, the variable `VERSION_IS_RELEASE` is not defined for the user as opposed to `SFML_FOUND` or `SFML_VERSION` for example. I would expect it to be available so it can be checked by the user, and it is documented as a bool output variable. Not sure how to make available though.

**Comment by kimci86** [comment]
Well, it can be done as simply as `set(VERSION_IS_RELEASE @VERSION_IS_RELEASE@)` in SFMLConfig.cmake.in.
While we're at it, what about adding "SFML_" as a prefix to this output variable?

**Comment by eXpl0it3r** [comment]
Should it be stored in the cache or not? 

For example:

image

```cmake
        set(SFML_VERSION_IS_RELEASE @VERSION_IS_RELEASE CACHE BOOL "Found SFML version is an official release")
        mark_as_advanced(SFML_VERSION_IS_RELEASE)
```

vs

```cmake
        set(SFML_VERSION_IS_RELEASE @VERSION_IS_RELEASE)
```

**Comment by eXpl0it3r** [comment]
I guess it has to be cached, otherwise you can't use it

**Comment by kimci86** [comment]
I am not very familiar with CMake technicalities, but I tried this locally and it works as expected.

In SFMLConfig.cmake.in:
```cmake
set(SFML_VERSION_IS_RELEASE @VERSION_IS_RELEASE@)
```

In the CMakeLists.txt of an application using SFML:
```cmake
find_package(SFML 3.0 COMPONENTS graphics audio REQUIRED)

if(NOT SFML_VERSION_IS_RELEASE)
	message(FATAL_ERROR "Oh no! I want stable API")
endif()
```
I don't know if putting the variable in cache makes sense.

**Comment by eXpl0it3r** [comment]
Yeah, just tested this myself as well.
The cache doesn't seem to make sense to me either, because it isn't something the user should be able to change.

**Pull request merged**
